### PR TITLE
Adding #issuecomment to the URL in E-Mail notifications

### PR DESCRIPTION
**Pull request by jonasfranz**
Reference: #1669

### CODE REVIEW

File: models/mail.go [hunk 1]
```diff
@@ -148,6 +148,24 @@ func composeTplData(subject, body, link string) map[string]interface{} {
 	return data
 }
 
+func composeIssueCommentMessage(issue *Issue, doer *User, comment *Comment, tplName base.TplName, tos []string, info string) *mailer.Message {
+	subject := issue.mailSubject()
+	body := string(markdown.RenderString(issue.Content, issue.Repo.HTMLURL(), issue.Repo.ComposeMetas()))
+	data := composeTplData(subject, body, issue.HTMLURL() + "#" +  comment.HashTag())
```
Comment (bkcsoft) [inline models/mail.go]: There's `comment.HTMLURL()` to replace `issue.HTMLURL() + "#" + comment.HashTag()` ... Also `go-fmt` :)

**Review by bkcsoft — CHANGES_REQUESTED** [review]

File: models/mail.go (hunk 1 above)
Comment (bkcsoft) [reply models/mail.go]: If you'd like, you can change `comment.HTMLURL()` to do this instead 😄 
```
	return fmt.Sprintf("%s#%s", issue.HTMLURL(), c.HashTag())
```

File: models/issue_mail.go [hunk 2]
```diff
@@ -18,22 +18,76 @@ func (issue *Issue) mailSubject() string {
...
-func mailIssueCommentToParticipants(issue *Issue, doer *User, mentions []string) error {
+func mailIssueCommentToParticipants(issue *Issue, doer *User, comment *Comment, mentions []string) error {
+
+	names, tos, err := prepareMailToParticipants(issue, doer)
+
+	if err != nil {
+		return fmt.Errorf("PrepareMailToParticipants: %v", err)
+	}
+
+	SendIssueCommentMail(issue, doer, comment, tos)
```
Comment (bkcsoft) [inline models/issue_mail.go]: Couldn't we merge these 2 (almost identical) functions by doing
```go
if comment != nil {
  SendIssueCommentMail(...)
} else {
  SendIssueActionMail(...)
}
```

File: models/issue_mail.go [hunk 3]
```diff
@@ -18,22 +18,76 @@ func (issue *Issue) mailSubject() string {
...
 }
 
-// mailIssueCommentToParticipants can be used for both new issue creation and comment.
+// mailIssueCommentToParticipants can be used only for comment.
 // This function sends two list of emails:
 // 1. Repository watchers and users who are participated in comments.
 // 2. Users who are not in 1. but get mentioned in current issue/comment.
-func mailIssueCommentToParticipants(issue *Issue, doer *User, mentions []string) error {
+func mailIssueCommentToParticipants(issue *Issue, doer *User, comment *Comment, mentions []string) error {
+
```
Comment (bkcsoft) [inline models/issue_mail.go]: no newline

File: models/issue_mail.go [hunk 4]
```diff
@@ -18,22 +18,76 @@ func (issue *Issue) mailSubject() string {
...
+	names, tos, err := prepareMailToParticipants(issue, doer)
+
+	if err != nil {
```
Comment (bkcsoft) [inline models/issue_mail.go]: no newline between `set and check`s :)

File: models/issue_mail.go [hunk 5]
```diff
@@ -18,22 +18,76 @@ func (issue *Issue) mailSubject() string {
...
+func prepareMailToParticipants(issue *Issue, doer *User) (tos []string, names []string, error error) {
 	if !setting.Service.EnableNotifyMail {
-		return nil
+		return nil, nil, nil
 	}
 
 	watchers, err := GetWatchers(issue.RepoID)
 	if err != nil {
-		return fmt.Errorf("GetWatchers [repo_id: %d]: %v", issue.RepoID, err)
+		return nil, nil, err
```
Comment (bkcsoft) [inline models/issue_mail.go]: the error should still we wrapped:
```go
  return nil, nil, fmt.Errorf("GetWatchers [repo_id: %d]: %v", issue.RepoID, err)
```

File: models/issue_mail.go [hunk 6]
```diff
@@ -18,22 +18,76 @@ func (issue *Issue) mailSubject() string {
...
 	}
 	participants, err := GetParticipantsByIssueID(issue.ID)
 	if err != nil {
-		return fmt.Errorf("GetParticipantsByIssueID [issue_id: %d]: %v", issue.ID, err)
+		return nil, nil, err
```
Comment (bkcsoft) [inline models/issue_mail.go]: Same here, wrap the error :)

File: models/issue_mail.go [hunk 7]
```diff
@@ -18,22 +18,76 @@ func (issue *Issue) mailSubject() string {
...
+
+		tos = append(tos, mentions[i])
+	}
+	SendIssueMentionInActionMail(issue, doer, GetUserEmailsByNames(tos))
+
+	return nil
+}
+
+// prepareMailToParticipants creates the tos and names list for an issue and the issue's creator.
+func prepareMailToParticipants(issue *Issue, doer *User) (tos []string, names []string, error error) {
```
Comment (bkcsoft) [inline models/issue_mail.go]: I'm not really a fan for named returns, got a reason to do that?

File: models/issue_mail.go [hunk 8]
```diff
@@ -42,16 +96,16 @@ func mailIssueCommentToParticipants(issue *Issue, doer *User, mentions []string)
...
+	names = make([]string, 0, len(watchers))
 	for i := range watchers {
 		if watchers[i].UserID == doer.ID {
 			continue
 		}
 
 		to, err := GetUserByID(watchers[i].UserID)
 		if err != nil {
-			return fmt.Errorf("GetUserByID [%d]: %v", watchers[i].UserID, err)
+			return nil, nil, err
```
Comment (bkcsoft) [inline models/issue_mail.go]: Wrap error

File: models/issue_mail.go [hunk 9]
```diff
@@ -95,7 +135,7 @@ func (issue *Issue) MailParticipants() (err error) {
 		return fmt.Errorf("UpdateIssueMentions [%d]: %v", issue.ID, err)
 	}
 
-	if err = mailIssueCommentToParticipants(issue, issue.Poster, mentions); err != nil {
+	if err = mailIssueActionToParticipants(issue, issue.Poster, mentions); err != nil {
```
Comment (bkcsoft) [inline models/issue_mail.go]: here you'd just call `mailIssueCommentToParticipants(issue, issue.Poster, nil, mentions)` instead

File: models/issue_mail.go [hunk 10]
```diff
@@ -95,7 +135,7 @@ func (issue *Issue) MailParticipants() (err error) {
...
 		log.Error(4, "mailIssueCommentToParticipants: %v", err)
```
Comment (bkcsoft) [inline models/issue_mail.go]: technically wrong here, but maybe not later on ;)

**Review by bkcsoft — CHANGES_REQUESTED** [review]
All the changes :trollface:

File: models/issue_mail.go [hunk 11]
```diff
@@ -95,7 +104,7 @@ func (issue *Issue) MailParticipants() (err error) {
...
```
Comment (bkcsoft) [inline models/issue_mail.go]: Wrong error-message, otherwise LGTM 🎉

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Review by bkcsoft — APPROVED** [review]

**Comment by bkcsoft** [comment]
LGTM, but untested

File: models/issue_mail.go [hunk 12]
```diff
@@ -18,11 +18,11 @@ func (issue *Issue) mailSubject() string {
 	return fmt.Sprintf("[%s] %s (#%d)", issue.Repo.Name, issue.Title, issue.Index)
 }
 
-// mailIssueCommentToParticipants can be used for both new issue creation and comment.
+// mailIssueCommentToParticipants can be used only for comment.
```
Comment (lafriks) [inline models/issue_mail.go]: Shouldn't this comment be changed back to original?

File: models/mail.go [hunk 13]
```diff
@@ -148,10 +148,16 @@ func composeTplData(subject, body, link string) map[string]interface{} {
...
 
-func composeIssueMessage(issue *Issue, doer *User, tplName base.TplName, tos []string, info string) *mailer.Message {
+func composeIssueCommentMessage(issue *Issue, doer *User, comment *Comment, tplName base.TplName, tos []string, info string) *mailer.Message {
 	subject := issue.mailSubject()
 	body := string(markdown.RenderString(issue.Content, issue.Repo.HTMLURL(), issue.Repo.ComposeMetas()))
-	data := composeTplData(subject, body, issue.HTMLURL())
+
+	data := make(map[string]interface{}, 10)
+	if comment != nil {
+		data = composeTplData(subject, body, issue.HTMLURL()+"#"+comment.HashTag())
```
Comment (lafriks) [inline models/mail.go:157]: Why comment.HTMLURL is not used here?

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/mail.go (hunk 13 above)
Comment (jonasfranz) [reply models/mail.go:157]: No because comment.HTMLURL returns an empty string by printing an error to the console (see gitter chat).

Comment (lafriks) [reply models/mail.go:157]: Ok

**Review by lafriks — APPROVED** [review]
Looks good but have to test it

File: models/mail.go (hunk 13 above)
Comment (bkcsoft) [reply models/mail.go:157]: @JonasFranzDEV Gitter chat is not searchable, mind sharing with the rest of the class? As I'm quite sure there's a better solution to this than re-implementing the wheel 🙂

**Comment by lunny** [comment]
LGTM

**Pull request merged**